core/amber/src/main/scala/edu/uci/ics/amber/engine/common/AmberConfig.scala

```diff
@@ -105,4 +108,66 @@ object AmberConfig {
   // JDBC configuration
   val jdbcConfig: Config = getConfSource.getConfig("jdbc")
 
+  // Language server configuration
+  val languageServer: String = getConfSource.getString("language-server.default")
+
+  private val logger: Logger = Logger.getLogger(this.getClass.getName)
+
+  //free the port for the server
+  private def releasePort(port: Int): Unit = {
+    val pythonCode =
+      s"""
+         |import os
+         |
+         |def release_port(port):
+         |    cmd_find = 'netstat -aon | findstr %s' % (port)
+         |    print('cmd_find:', cmd_find)
+         |    res = os.popen(cmd_find).read()
+         |    print('res:', res)
+         |    if str(port) and 'LISTENING' in res:
+         |        i = res.index('LISTENING')
+         |        start = i + len('LISTENING') + 7
+         |        end = res.find('\\n', start)
+         |        pid = res[start:end].strip()
+         |        print('pid:', pid)
+         |        cmd_kill = 'taskkill -f -pid %s' % (pid)
+         |        print('cmd_kill:', cmd_kill)
+         |        os.popen(cmd_kill)
+         |    else:
+         |        print('This port is free to use')
+         |
+         |release_port($port)
+       """.stripMargin
+
+    val command = Seq("python", "-c", pythonCode)
+    val exitCode = command.!
+    if (exitCode == 0) {
+      logger.info(s"Successfully free the port: $port")
+    } else {
+      logger.warning(s"fail to free the port: $port")
+    }
+  }
+
+  def startLanguageServer(): Unit = {
+    languageServer match {
+      case "pyright" =>
+        logger.info("Starting Pyright...")
+        releasePort(3000)
+        //Try to run the Pyright language server
+        try {
+          val result =
+            Process("node ../languageServer/startPyright.mjs").run()
+          logger.info("Pyright language server is running on port 3000")
+        } catch {
+          case e: Exception => logger.warning(s"Failed to start Pyright: ${e.getMessage}")
```

Comment:
Do you have any sugession for the retry logic? I use a `release-port` function to avoid the occupied port situation, and I havan't meet any other error situation.

Reply:
we have retry logic in this repo at places we need to assign ports for proxy servers. The point is that your python code tries to acquire a port and then release it, then immediately use the port for starting a socket. In rare (but it could happen) cases, the port can be acquired by another process right after you release it, and becomes unavailable when you try to use it with the socket. And the socket will fail to start. That's why this logic needs a retry process.

Reply:
I see. I have add the retry logic.